src/artemis/external_interaction/zocalo_interaction.py

```diff
@@ -1,114 +1,184 @@
-        else:
-            artemis.log.LOGGER.warn(
-                f"Warning: results for {received_group_id} received but expected \
-                    {data_collection_group_id}"
-            )
-
-    workflows.recipe.wrap_subscribe(
-        transport,
-        "xrc.i03",
-        receive_result,
-        acknowledgement=True,
-        allow_non_recipe_messages=False,
-    )
-
-    try:
-        return result_received.get(timeout=timeout)
-    finally:
-        transport.disconnect()
+    def __init__(self, parameters: FullParameters, ispyb_handler: ISPyBHandlerCallback):
+        self.grid_position_to_motor_position = (
+            parameters.grid_scan_params.grid_position_to_motor_position
+        )
+        self.zocalo_env = parameters.zocalo_environment
+        self.processing_start_time = 0.0
+        self.processing_time = 0.0
+        self.results = None
+        self.xray_centre_motor_position = None
+        self.ispyb = ispyb_handler
+
+    def event(self, doc: dict):
+        LOGGER.debug(f"\n\nZocalo handler received event document:\n\n {doc}\n")
+        descriptor = self.ispyb.descriptors.get(doc["descriptor"])
+        assert descriptor is not None
+        event_name = descriptor.get("name")
+        if event_name == ISPYB_PLAN_NAME:
+            if self.ispyb.ispyb_ids[0] is not None:
+                datacollection_ids = self.ispyb.ispyb_ids[0]
+                for id in datacollection_ids:
+                    self.run_start(id)
+            else:
+                raise ISPyBDepositionNotMade("ISPyB deposition was not initialised!")
+
+    def stop(self, doc: dict):
+        LOGGER.debug(f"\n\nZocalo handler received stop document:\n\n {doc}\n")
+        if self.ispyb.ispyb_ids == (None, None, None):
+            raise ISPyBDepositionNotMade("ISPyB deposition was not initialised!")
+        datacollection_ids = self.ispyb.ispyb_ids[0]
+        for id in datacollection_ids:
+            self.run_end(id)
+        self.processing_start_time = time.time()
+
+    def wait_for_results(self):
+        datacollection_group_id = self.ispyb.ispyb_ids[2]
+        raw_results = self.wait_for_result(datacollection_group_id)
+        self.processing_time = time.time() - self.processing_start_time
+        # wait_for_result returns the centre of the grid box, but we want the corner
+        self.results = Point3D(
+            raw_results.x - 0.5, raw_results.y - 0.5, raw_results.z - 0.5
+        )
+        self.xray_centre_motor_position = self.grid_position_to_motor_position(
+            self.results
+        )
+
+        LOGGER.info(f"Results recieved from zocalo: {self.xray_centre_motor_position}")
+        LOGGER.info(f"Zocalo processing took {self.processing_time}s")
+
+    def _get_zocalo_connection(self, env: str = "artemis"):
+        zc = zocalo.configuration.from_file()
+        zc.activate_environment(env)
+
+        transport = lookup("PikaTransport")()
+        transport.connect()
+
+        return transport
```

Comment:
Could: Now we're holding state it might make sense to just create the connection once on init and reuse it each time?